src/vis/violin/utils.ts

```diff
@@ -0,0 +1,142 @@
+import {merge} from 'lodash';
+import {CategoricalColumn, ColumnInfo, EColumnTypes, ESupportedPlotlyVis, IVisConfig, NumericalColumn, Scales} from '../interfaces';
+import {PlotlyInfo, PlotlyData} from '../interfaces';
+import {EViolinOverlay} from '../bar/utils';
+
+export function isViolin(s: IVisConfig): s is IViolinConfig {
+    return s.type === ESupportedPlotlyVis.VIOLIN;
+}
+
+export interface IViolinConfig {
+    type: ESupportedPlotlyVis.VIOLIN;
+    numColumnsSelected: ColumnInfo[];
+    catColumnsSelected: ColumnInfo[];
+    violinOverlay: EViolinOverlay;
+}
+
+const defaultConfig: IViolinConfig = {
+    type: ESupportedPlotlyVis.VIOLIN,
+    numColumnsSelected: [],
+    catColumnsSelected: [],
+    violinOverlay: EViolinOverlay.NONE,
+};
+
+export function violinMergeDefaultConfig(
+    columns: (NumericalColumn | CategoricalColumn)[],
+    config: IViolinConfig,
+): IVisConfig {
+    const merged = merge({}, defaultConfig, config);
+
+    const numCols = columns.filter((c) => c.type === EColumnTypes.NUMERICAL);
+
+    if(merged.numColumnsSelected.length === 0 && numCols.length > 0) {
+        merged.numColumnsSelected.push(numCols[numCols.length - 1].info);
+    }
+
+    return merged;
+}
+
+export function createViolinTraces(
+    columns: (NumericalColumn | CategoricalColumn)[],
+    config: IViolinConfig,
+    scales: Scales,
+): PlotlyInfo {
+    let counter = 1;
+
+    if(!config.numColumnsSelected || !config.catColumnsSelected) {
+        return {
+            plots: [],
+            legendPlots: [],
+            rows: 0,
+            cols: 0,
+            errorMessage: 'To create a Violin plot, please select at least 1 numerical column.',
+            formList: ['violinOverlay']
+        };
+    }
+
+    const numCols: NumericalColumn[] = columns.filter((c) => config.numColumnsSelected.filter((d) => c.info.id === d.id).length > 0 && EColumnTypes.NUMERICAL) as NumericalColumn[];
```

Comment:
I am not sure if I understand the filter condition (also the one below), i.e. what does the `&& EColumnTypes.NUMERICAL` do? Shouldn't there be a `&& c.something === EColumnTypes.NUMERICAL` instead?

Reply:
I just found it: it should probably look like:
```typescript
const numCols: NumericalColumn[] = columns.filter((c) => config.numColumnsSelected.filter((d) => c.info.id === d.id).length > 0 && c.type === EColumnTypes.NUMERICAL) as NumericalColumn[];
```

Also, instead of checking the length if you have any matches, use `.some(` as it returns early (i.e. when it found a single match):

```typescript
const numCols: NumericalColumn[] = columns.filter((c) => config.numColumnsSelected.some((d) => c.info.id === d.id) && c.type === EColumnTypes.NUMERICAL) as NumericalColumn[];
```

Reply:
good catch, yes that was a bug. and the some feature is great to know about! Changed a few other places to also use .some